File: server/mc/1.21.1-neoforge/src/main/java/calebxzhou/rdi/mc/server/rcmd/RcmdServerCommands.java
package calebxzhou.rdi.mc.server.rcmd;

import calebxzhou.rdi.mc.rcmd.RcmdArgumentTypes;
import calebxzhou.rdi.mc.rcmd.RcmdCommandSpec;
import calebxzhou.rdi.mc.rcmd.RcmdContext;
import calebxzhou.rdi.mc.rcmd.RcmdDispatcher;
import calebxzhou.rdi.mc.rcmd.RcmdResult;
import calebxzhou.rdi.mc.rcmd.RcmdSource;
import calebxzhou.rdi.mc.common2.chat.ChatRange;
import calebxzhou.rdi.mc.common2.chat.PlayerChatRangeState;
import calebxzhou.rdi.mc.common2.home.HomeResult;
import calebxzhou.rdi.mc.common2.home.HomeService;
import calebxzhou.rdi.mc.common2.tpa.TpaResult;
import calebxzhou.rdi.mc.common2.tpa.TpaService;
import calebxzhou.rdi.mc.server.home.HomePlayer211;
import calebxzhou.rdi.mc.server.tpa.TpaPlayer211;
import calebxzhou.rdi.mc.server.tpa.TpaPlayerLookup211;
import net.minecraft.server.level.ServerPlayer;

public final class RcmdServerCommands {
    private static final RcmdDispatcher DISPATCHER = new RcmdDispatcher();

    static {
        DISPATCHER.register(
                RcmdCommandSpec.builder("ping")
                        .description("Test rcmd availability")
                        .command(context -> RcmdResult.ok("pong，rcmd正常"))
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("chat", "range")
                        .description("Set chat range")
                        .argument("range", RcmdArgumentTypes.enumOf("host", "global"))
                        .command(context -> {
                            var range = context.getString("range");
                            var source = context.source();
                            if (!source.isPlayer()) {
                                return RcmdResult.error("此rcmd命令只能由玩家执行");
                            }
                            var chatRange = ChatRange.fromRcmdValue(range);
                            PlayerChatRangeState.set(source.playerId(), chatRange);
                            return RcmdResult.ok("聊天范围已切换为" + chatRange.getDisplayName());
                        })
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("tpa")
                        .description("Request teleport to another player")
                        .argument("playerName", RcmdArgumentTypes.STRING)
                        .command(RcmdServerCommands::handleTpa)
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("tpok")
                        .description("Accept pending teleport request")
                        .command(RcmdServerCommands::handleTpok)
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("sethome")
                        .description("Save current player position as a home")
                        .argument("name", RcmdArgumentTypes.STRING)
                        .command(RcmdServerCommands::handleSetHome)
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("home")
                        .description("Teleport player to a saved home")
                        .argument("name", RcmdArgumentTypes.STRING)
                        .command(RcmdServerCommands::handleHome)
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("listhome")
                        .description("List saved player homes")
                        .command(RcmdServerCommands::handleListHome)
                        .build()
        );
        DISPATCHER.register(
                RcmdCommandSpec.builder("delhome")
                        .description("Delete a saved player home")
                        .argument("name", RcmdArgumentTypes.STRING)
                        .command(RcmdServerCommands::handleDeleteHome)
                        .build()
        );
    }

    private RcmdServerCommands() {
    }

    public static RcmdDispatcher dispatcher() {
        return DISPATCHER;
    }

    public static void reply(RcmdSource source, RcmdResult result) {
        if (result == null || result.message().isEmpty()) {
            return;
        }
        for (var message : result.message().split("\\R")) {
            replyLine(source, result.success(), message);
        }
    }

    private static void replyLine(RcmdSource source, boolean success, String message) {
        if (message.isEmpty()) {
            return;
        }
        switch (success ? ReplyKind.SUCCESS : ReplyKind.ERROR) {
            case SUCCESS -> source.sendFeedback(message);
            case ERROR -> source.sendError(message);
        }
    }

    public static String getChatRange(java.util.UUID playerId) {
        return PlayerChatRangeState.get(playerId).name().toLowerCase();
    }

    private static RcmdResult handleTpa(RcmdContext context) {
        ServerPlayer requester = playerOrNull(context.source());
        if (requester == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var targetName = context.getString("playerName");
        var result = TpaService.request(new TpaPlayer211(requester), targetName, new TpaPlayerLookup211(requester.server));
        return toRcmdResult(result);
    }

    private static RcmdResult handleTpok(RcmdContext context) {
        ServerPlayer target = playerOrNull(context.source());
        if (target == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var result = TpaService.accept(new TpaPlayer211(target), new TpaPlayerLookup211(target.server));
        return toRcmdResult(result);
    }

    private static RcmdResult handleSetHome(RcmdContext context) {
        ServerPlayer player = playerOrNull(context.source());
        if (player == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var result = HomeService.setHome(new HomePlayer211(player), context.getString("name"));
        return toRcmdResult(result);
    }

    private static RcmdResult handleHome(RcmdContext context) {
        ServerPlayer player = playerOrNull(context.source());
        if (player == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var result = HomeService.goHome(new HomePlayer211(player), context.getString("name"));
        return toRcmdResult(result);
    }

    private static RcmdResult handleListHome(RcmdContext context) {
        ServerPlayer player = playerOrNull(context.source());
        if (player == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var result = HomeService.listHomes(new HomePlayer211(player));
        return toRcmdResult(result);
    }

    private static RcmdResult handleDeleteHome(RcmdContext context) {
        ServerPlayer player = playerOrNull(context.source());
        if (player == null) {
            return RcmdResult.error("此rcmd命令只能由玩家执行");
        }
        var result = HomeService.deleteHome(new HomePlayer211(player), context.getString("name"));
        return toRcmdResult(result);
    }

    private static RcmdResult toRcmdResult(TpaResult result) {
        return result.success() ? RcmdResult.ok(result.message()) : RcmdResult.error(result.message());
    }

    private static RcmdResult toRcmdResult(HomeResult result) {
        return result.success() ? RcmdResult.ok(result.message()) : RcmdResult.error(result.message());
    }

    private static ServerPlayer playerOrNull(RcmdSource source) {
        if (source instanceof RcmdServerSource211 playerSource) {
            return playerSource.getPlayer();
        }
        if (source instanceof RcmdCommandSourceStackSource commandSource) {
            return commandSource.getPlayer();
        }
        return null;
    }

    private enum ReplyKind {
        SUCCESS,
        ERROR
    }
}

File: server/mc/1.21.1-neoforge/src/main/kotlin/calebxzhou/rdi/mc/server/rcmd/RcmdServerCommands.kt
package calebxzhou.rdi.mc.server.rcmd

import calebxzhou.rdi.mc.common.RDI
import calebxzhou.rdi.mc.common3.mcs
import calebxzhou.rdi.mc.firmsection.FirmSectionKey
import calebxzhou.rdi.mc.firmsection.FirmSectionLimits
import calebxzhou.rdi.mc.firmsection.FirmSectionSetStatus
import calebxzhou.rdi.mc.rcmd.*
import calebxzhou.rdi.mc.rcmd.chat.ChatRange
import calebxzhou.rdi.mc.rcmd.chat.PlayerChatRangeState
import calebxzhou.rdi.mc.rcmd.home.HomeResult
import calebxzhou.rdi.mc.rcmd.home.HomeService
import calebxzhou.rdi.mc.rcmd.tpa.TpaResult
import calebxzhou.rdi.mc.rcmd.tpa.TpaService
import calebxzhou.rdi.mc.server.firmsection.FirmSectionService
import net.minecraft.resources.ResourceKey
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.event.entity.player.PlayerEvent.PlayerLoggedOutEvent
import net.neoforged.neoforge.event.tick.PlayerTickEvent
import java.util.Locale
import java.util.UUID

@EventBusSubscriber(modid = "rdi")
object RcmdServerCommands : RcmdServerCommandHandler {
    private val DISPATCHER = RcmdDispatcher()
    private val POS_LOCKS = mutableMapOf<UUID, PosLockState>()
    private const val POS_LOCK_MAX_DISTANCE_SQR = 0.0001
    private const val TEST_ENTITY_COUNT = 65535
    private const val TEST_ENTITY_ITEM_COUNT = 32

    init {
        RcmdCommonServerCommands.register(DISPATCHER, this, RDI.DEBUG)
    }

    @JvmStatic
    fun dispatcher(): RcmdDispatcher = DISPATCHER

    @JvmStatic
    fun reply(source: RcmdSource, result: RcmdResult?) {
        if (result == null || result.message.isEmpty()) {
            return
        }
        result.message
            .lineSequence()
            .filter { it.isNotEmpty() }
            .forEach { replyLine(source, result.success, it) }
    }

    private fun replyLine(source: RcmdSource, success: Boolean, message: String) {
        if (success) {
            source.sendFeedback(message)
        } else {
            source.sendError(message)
        }
    }

    fun getChatRange(playerId: UUID): String {
        return PlayerChatRangeState.get(playerId).name.lowercase(Locale.ROOT)
    }

    @SubscribeEvent
    @JvmStatic
    fun onPlayerTick(event: PlayerTickEvent.Post) {
        val player = event.entity as? ServerPlayer ?: return
        val state = POS_LOCKS[player.uuid] ?: return
        keepPosLocked(player, state)
    }

    @SubscribeEvent
    @JvmStatic
    fun onPlayerLogout(event: PlayerLoggedOutEvent) {
        val player = event.entity
        if (player is ServerPlayer) {
            val state = POS_LOCKS.remove(player.uuid)
            if (state != null) {
                restorePosLockState(player, state)
            }
        }
    }

    override fun ping(context: RcmdContext): RcmdResult = RcmdResult.ok("pong，rcmd正常")

    override fun setChatRange(context: RcmdContext): RcmdResult {
        val range = context.getString("range")
        val source = context.source
        if (!source.isPlayer) {
            return RcmdResult.error("此rcmd命令只能由玩家执行")
        }
        val chatRange = ChatRange.fromRcmdValue(range)
        PlayerChatRangeState.set(source.playerId(), chatRange)
        return RcmdResult.ok("聊天范围已切换为" + chatRange.displayName)
    }

    override fun togglePosLock(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val playerId = player.uuid
        val removed = POS_LOCKS.remove(playerId)
        if (removed != null) {
            restorePosLockState(player, removed)
            return RcmdResult.ok("位置锁定已关闭")
        }
        POS_LOCKS[playerId] = PosLockState.from(player)
        player.isInvulnerable = true
        player.isInvisible = true
        stopPlayerMovement(player)
        return RcmdResult.ok("位置锁定已开启")
    }

    override fun setFirmSection(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = FirmSectionService.set(player)
        val label = firmSectionLabel(result.key)
        return when (result.status) {
            FirmSectionSetStatus.ADDED ->
                RcmdResult.ok("已持久当前子区块：$label ${firmSectionCountLabel(result.playerCount, result.total)}")

            FirmSectionSetStatus.ALREADY_PRESENT ->
                RcmdResult.ok("当前子区块已经持久了")

            FirmSectionSetStatus.PLAYER_LIMIT_REACHED ->
                RcmdResult.error("你持久的子区块已达到个人上限${FirmSectionLimits.maxPerson}个")

            FirmSectionSetStatus.TOTAL_LIMIT_REACHED ->
                RcmdResult.error("持久子区块已达到全世界上限${FirmSectionLimits.maxTotal}个")
        }
    }

    override fun unsetFirmSection(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = FirmSectionService.unset(player)
        val label = firmSectionLabel(result.key)
        return if (result.removed) {
            RcmdResult.ok("已取消持久当前子区块：$label ${firmSectionCountLabel(result.playerCount, result.total)}")
        } else {
            RcmdResult.ok("当前子区块尚未持久")
        }
    }

    override fun listFirmSections(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = FirmSectionService.list(player)
        if (result.sections.isEmpty()) {
            return RcmdResult.ok("你还没有持久子区块。${firmSectionCountLabel(result.playerCount, result.total)}")
        }
        val lines = buildList {
            add("持久子区块数量：${firmSectionCountLabel(result.playerCount, result.total)}")
            result.sections.groupBy { it.dimensionId }.forEach { (dimensionId, sections) ->
                add("$dimensionId : ${sections.map { firmSectionPositionLabel(it) }}")
            }
        }
        return RcmdResult.ok(lines.joinToString("\n"))
    }

    override fun setFirmSectionAutoSet(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val enabled = context.getBool("enabled")
        FirmSectionService.setAutoSetEnabled(player, enabled)
        return RcmdResult.ok("放置方块实体时自动设置持久子区块已${if (enabled) "开启" else "关闭"}")
    }

    override fun requestTpa(context: RcmdContext): RcmdResult {
        val requester = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val targetName = context.getString("playerName")
        val result = TpaService.request(TpaPlayer211(requester), targetName, TpaPlayerLookup211(requester.server))
        return toRcmdResult(result)
    }

    override fun acceptTpa(context: RcmdContext): RcmdResult {
        val target = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = TpaService.accept(TpaPlayer211(target), TpaPlayerLookup211(target.server))
        return toRcmdResult(result)
    }

    override fun setHome(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = HomeService.setHome(HomePlayer211(player), context.getString("name"))
        return toRcmdResult(result)
    }

    override fun goHome(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = HomeService.goHome(HomePlayer211(player), context.getString("name"))
        return toRcmdResult(result)
    }

    override fun listHome(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = HomeService.listHomes(HomePlayer211(player))
        return toRcmdResult(result)
    }

    override fun deleteHome(context: RcmdContext): RcmdResult {
        val player = playerOrNull(context.source) ?: return RcmdResult.error("此rcmd命令只能由玩家执行")
        val result = HomeService.deleteHome(HomePlayer211(player), context.getString("name"))
        return toRcmdResult(result)
    }

    override fun testEntity(context: RcmdContext): RcmdResult {
        mcs.execute {


            val player = playerOrNull(context.source) ?: return@execute
            val level = player.serverLevel()
            val x = player.x
            val y = player.y - 3.0
            val z = player.z
            var added = 0
            repeat(TEST_ENTITY_COUNT) {
                val itemEntity = ItemEntity(level, x, y, z, ItemStack(Items.DIAMOND_AXE, TEST_ENTITY_ITEM_COUNT))
                if (level.addFreshEntity(itemEntity)) {
                    added++
                }
            }
        }
        return RcmdResult.ok()
    }

    private fun toRcmdResult(result: TpaResult): RcmdResult =
        if (result.success) RcmdResult.ok(result.message) else RcmdResult.error(result.message)

    private fun toRcmdResult(result: HomeResult): RcmdResult =
        if (result.success) RcmdResult.ok(result.message) else RcmdResult.error(result.message)

    private fun keepPosLocked(player: ServerPlayer, state: PosLockState) {
        player.setInvulnerable(true)
        player.setInvisible(true)
        stopPlayerMovement(player)
        val targetLevel = player.server.getLevel(state.dimension)
        if (targetLevel == null) {
            restorePosLockState(player, state)
            POS_LOCKS.remove(player.uuid)
            return
        }
        val currentPos = player.position()
        val moved = currentPos.distanceToSqr(state.x, state.y, state.z) > POS_LOCK_MAX_DISTANCE_SQR
        if (moved || player.level().dimension() != state.dimension) {
            player.teleportTo(targetLevel, state.x, state.y, state.z, state.yaw, state.pitch)
            stopPlayerMovement(player)
        }
    }

    private fun restorePosLockState(player: ServerPlayer, state: PosLockState) {
        player.setInvulnerable(state.wasInvulnerable)
        player.setInvisible(state.wasInvisible)
        stopPlayerMovement(player)
    }

    private fun stopPlayerMovement(player: ServerPlayer) {
        player.setDeltaMovement(Vec3.ZERO)
        player.resetFallDistance()
    }

    private fun playerOrNull(source: RcmdSource): ServerPlayer? =
        when (source) {
            is RcmdServerSource211 -> source.player
            is RcmdCommandSourceStackSource -> source.player
            else -> null
        }

    private fun firmSectionLabel(key: FirmSectionKey): String =
        "${key.dimensionId},${key.chunkX},${key.sectionY},${key.chunkZ}"

    private fun firmSectionPositionLabel(key: FirmSectionKey): String =
        "${key.chunkX},${key.sectionY},${key.chunkZ}"

    private fun firmSectionCountLabel(playerCount: Int, total: Int): String =
        if (FirmSectionLimits.maxPerson > 0) {
            "你：${playerCount}/${FirmSectionLimits.maxPerson}，全世界：${total}/${FirmSectionLimits.maxTotal}"
        } else {
            "你：${playerCount}个，全世界：${total}/${FirmSectionLimits.maxTotal}"
        }

    @JvmRecord
    data class PosLockState(
        val dimension: ResourceKey<Level>,
        val x: Double,
        val y: Double,
        val z: Double,
        val yaw: Float,
        val pitch: Float,
        val wasInvulnerable: Boolean,
        val wasInvisible: Boolean
    ) {
        companion object {
            fun from(player: ServerPlayer) = PosLockState(
                player.level().dimension(),
                player.x,
                player.y,
                player.z,
                player.yRot,
                player.xRot,
                player.isInvulnerable,
                player.isInvisible
            )
        }
    }
}
